modtools/fire-rate.lua

```diff
@@ -0,0 +1,445 @@
+  skill_mode = "basic",
+}
+---------------------------------------------------------------------
+
+function register_weapon(weapon_token, material_token, data)
+  local key = weapon_token
+
+  if ( material_token ~= nil ) then
+    key = key ..  separator .. material_token
+  end
+
+  if ( data == nil ) then
+    data = {}
+  end
+
+  local new_entry = {
+    fort_rate = data.fort_rate or 80,
+    adv_rate = data.adv_rate or 80,
+    fort_skill_factor = data.fort_skill_factor or 2.7,
+    adv_skill_factor = data.adv_skill_factor or 2.7,
+  }
+
+  -- Determine caps
+  if ( data.fort_cap ) then
+    new_entry.fort_cap = data.fort_cap
+  else -- default to half main rate
+    new_entry.fort_cap = math.floor(new_entry.fort_rate / 2 + 0.5) -- Half the standard rate, then round to nearest number
+  end
+
+  -- Accounting for the adventure hard-coded cap won't be done at this stage
+  if ( data.adv_cap ) then
+    new_entry.adv_cap = data.adv_cap
+  else
+    new_entry.adv_cap = math.floor(new_entry.adv_rate / 2 + 0.5)
+  end
+
+  weapon_data[key] = new_entry
+  return new_entry
+end
+
+function get_weapon_data(weapon_token, material_token)
+  local key = weapon_token
+
+  -- Prioritise getting a particular weapon + material combination first
+  if ( material_token ~= nil ) then
+    local key_with_mat = key .. separator .. material_token
+
+    if ( weapon_data[key_with_mat] ~= nil ) then
+      return weapon_data[key_with_mat]
+    end
+  end
+
+  -- Otherwise, attempt to find just the weapon's particular data
+  if ( weapon_data[key] ~= nil ) then
+    return weapon_data[key]
+  end
+
+  -- If we get here, there is no entry
+  -- Just so that all ranged weapons are consistently using this new system, we'll create one.
+  -- Only for the weapon, not any particular weapon + material combo.
+  return register_weapon(weapon_token, nil)
+end
+
+function get_weapon_fire_rate(weapon_token, material_token, game_mode)
+  local weapon_data = get_weapon_data(weapon_token, material_token)
+
+  if ( game_mode == nil or game_mode == df.game_mode.DWARF ) then
+    return weapon_data.fort_rate
+  else -- assume df.game_mode.ADVENTURE
+    return weapon_data.adv_rate
+  end
+end
+
+function get_weapon_skill_factor(weapon_token, material_token, game_mode)
+  local weapon_data = get_weapon_data(weapon_token, material_token)
+
+  if ( game_mode == nil or game_mode == df.game_mode.DWARF ) then
+    return weapon_data.fort_skill_factor
+  else -- assume ADVENTURE
+    return weapon_data.adv_skill_factor
+  end
+end
+
+
+function get_weapon_fire_rate_cap(weapon_token, material_token, game_mode)
+  local weapon_data = get_weapon_data(weapon_token, material_token)
+
+  if ( game_mode == nil or game_mode == df.game_mode.DWARF ) then
+    return weapon_data.fort_cap
+  else -- assume ADVENTURE
+    return weapon_data.adv_cap
+  end
+end
+
+function set_config_skill_mode(value)
+  if ( value ~= "basic" and value ~= "vanilla" ) then
+    config.skill_mode = "basic"
```

Comment:
This if block needs a return statement, otherwise the (invalid) value will be used a few lines down

Reply:
oops - fixed